### ISSUE TITLE
Apache Aurora Support

### ISSUE THREAD

**Issue by payomdousti** [deis/deis#3538]
From what I understand, Aurora addresses every issue that @gabrtv had with Marathon. Is there any specific reasons why Aurora isn't being considered at the moment?

This seems like it's the best way to start leveraging Mesos within Deis.

**Comment by gabrtv** [comment]
@payomdousti I agree Aurora is worth a deeper look. At this point, we are still investigating:

 * Mesos with an off-the-shelf framework (maybe Aurora?)
 * Mesos with a custom framework
 * Kubernetes (see #2744)
 * Docker Swarm (see #3134)

ping @smothiki

**Comment by payomdousti** [comment]
I've taken a look at Kubernetes and Docker Swarm in the context of Deis, and it seems while these may potentially become good products, they are lacking the maturity and feature set of Aurora. 

1) From a scheduling perspective Aurora is the only one that is resource aware at this point in time, which is a major win.

2) Aurora has been battle tested supporting the Twitter platform for the past couple of years. Alternatively Swarm is in alpha, and Kubernetes also seems relatively immature in comparison.

Maybe I'm the minority here, but aurora support would convince me to commit to deis as our PaaS moving forward.

**Comment by smothiki** [comment]
@payomdousti  I have looked at Aurora and felt like its much better than Marathon . At present Deis team is working on supporting swarm scheduler and helping swarm project. Swarm team is also coming up with Mesos integration with their custom framework . Most probably next week we will start working on integrating Mesos  and first goal will be creating a mock prototype with Aurora Framework ,test it for Deis requirements and decide whether to write a lightweight custom framework for Deis .

**Comment by payomdousti** [comment]
any updates on this? anything i can help out with?

**Comment by smothiki** [comment]
@payomdousti right now we are done with swarm preview and dealing with https://github.com/deis/deis/issues/3605. Once the issue is fixed we will jump on prototyping aurora ASAP late next week.

**Comment by benley** [comment]
Several people from the aurora-dev list are lurking on this bug and would be happy to collaborate, for what it's worth :)

**Comment by smothiki** [comment]
@payomdousti @benley we have started playing with aurora. Is there any api list for aurora framework something similar to [kubernetes](http://kubernetes.io/third_party/swagger-ui/#!/api/getApiVersions) . Having api endpoints and Json format for post calls makes it easier for deis to integrate with aurora. Right now all I can see is aurora python client and lots of coding from DEIS end to mock client. We need only [jobs] (https://github.com/apache/aurora/blob/master/src/main/python/apache/aurora/client/cli/jobs.py) fucntions of the client but to achieve that takes good amount of coding.

**Comment by bacongobbler** [comment]
@smothiki aurora implements a thrift RPC API, according to https://github.com/apache/aurora-website/blob/master/source/documentation/latest/developing-aurora-client.md#overview-of-the-client-architecture. I haven't found any from digging around, but I'd imagine that [this](https://github.com/apache/aurora/blob/master/api/src/main/thrift/org/apache/aurora/gen/api.thrift) would be the API endpoints.

**Comment by smothiki** [comment]
just came across this Note:( The only container type currently supported is “docker”. Docker support is currently EXPERIMENTAL.) in aurora [guide](http://aurora.apache.org/documentation/latest/configuration-reference/) also I haven't found any configuration guides or examples that specify environment variables , using volumes and port bind mounting . Will dig deep in the code and see if there is any other way to specify these setting in the configuration file.

**Comment by benley** [comment]
If you have a running aurora scheduler, http://your-scheduler/apiclient/api.html ought to give you a page explaining all the thrift api endpoints.

As for container types: What the docs actually mean there is the only container type **other than normal mesos containers** is "docker".  If you don't explicitly configure a docker image for an aurora job, it will use the mesos default.

Environment variables: thermos executor handles setting up the task runtime environment.  It actually doesn't have explicit hooks for setting environment variables, which still confuses me.  It does easily let you run arbitrary shell scripts to set up your task, and thermos sources a .thermos_profile file if it is present when starting processes, so you can simply do something like

```sh
cat > .thermos_profile <<EOF
export ENV_VAR=asdfasdf
export ENV_VAR2=asdfasdf
# etc
EOF

in an initialization process to set up the environment for further processes in the task.

Volumes and port mapping: I'm not sure aurora currently supports these; need to check the status of that.

**Comment by benley** [comment]
Other things that are a bit odd about aurora's docker support:

* It completely ignores any entrypoint, cmd, or environment variables that are specified by the image; thermos always starts with an empty environment.
* docker images need to have python preinstalled in order for the thermos executor to work in them.

I know there is work in progress to allow using Aurora with custom executors, so it should become possible to skip thermos entirely if desired.  It might make sense for Deis to have its own executor, depending on your requirements.

**Comment by mboersma** [comment]
Mesos + Marathon tech preview landed in master in #3612. :cake: 

Deis always welcomes a PR from the community, but adding Aurora scheduler support isn't on [the roadmap](http://docs.deis.io/en/latest/roadmap/roadmap/). Please do attend the August [open planning meeting](http://docs.deis.io/en/latest/roadmap/planning/#release-planning-meetings) to voice your support for Aurora if that's something you would like to see the project work on in the future.
